src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php

```diff
@@ -42,51 +42,72 @@ public function __construct(Connection $connection, SerializerInterface $seriali
     public function receive(callable $handler): void
     {
         while (!$this->shouldStop) {
-            $AMQPEnvelope = $this->connection->get();
+            try {
+                $AMQPEnvelope = $this->connection->get();
+            } catch (\AMQPException $exception) {
+                throw new TransportException($exception->getMessage(), 0, $exception);
+            }
+
             if (null === $AMQPEnvelope) {
-                $handler(null);
+                $handler(null, null);
 
-                usleep($this->connection->getConnectionCredentials()['loop_sleep'] ?? 200000);
+                usleep($this->connection->getConnectionConfiguration()['loop_sleep'] ?? 200000);
                 if (\function_exists('pcntl_signal_dispatch')) {
                     pcntl_signal_dispatch();
                 }
 
                 continue;
             }
 
-            try {
-                $handler($this->serializer->decode([
-                    'body' => $AMQPEnvelope->getBody(),
-                    'headers' => $AMQPEnvelope->getHeaders(),
-                ]));
-
-                $this->connection->ack($AMQPEnvelope);
-            } catch (RejectMessageExceptionInterface $e) {
-                try {
-                    $this->connection->reject($AMQPEnvelope);
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
-
-                throw $e;
-            } catch (\AMQPException $e) {
-                throw new TransportException($e->getMessage(), 0, $e);
-            } catch (\Throwable $e) {
-                try {
-                    $this->connection->nack($AMQPEnvelope, AMQP_REQUEUE);
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
+            $handler($this->serializer->decode([
+                'body' => $AMQPEnvelope->getBody(),
+                'headers' => $AMQPEnvelope->getHeaders(),
+            ]), new QueuedMessageMetadata(
```

Comment:
Can't this `QueuedMessageMetadata` simply be a Stamp added to the decoded envelope? It's very much related to retries so I'd make it explicit in the name.

Reply:
I've changed this to a stamp indeed! Thought, the stamp now only carries the current retry number, the "message" is carried via "state" inside the receiver now... which I did try avoid leaking this internal object externally on a stamp.